### ISSUE TITLE
`ref` can't take a symbol

### ISSUE THREAD

**Issue by ljharb** [rricard/proposal-refcollection#6]
If `ref` takes a symbol, then I can pass it a registered symbol (`Symbol.for('foo')`) or a well-known symbol (`Symbol.iterator`).

These can never be garbage collected, which is why Symbols are not permitted in WeakMap keys, or WeakRefs.

In other words, there's an established precedent that only unregistered symbols could be weakly held, and that any API that takes both kinds of symbols must treat them the same, which means no API that takes a symbol can weakly hold it. (cc @erights @devsnek).

**Comment by Jack-Works** [comment]
It doesn't make sense, you can store an object to global and make it "never be garbage collected" too.

**Comment by ljharb** [comment]
That is true, but that's still the reasoning that applies for Weak things not accepting symbols. See https://github.com/tc39/ecma262/issues/1194 for more discussion.

**Comment by erights** [comment]
The optional symbol second parameter to `ref` surprised me. Before we argue about whether it violates language constraints, I'd like to understand: What's it for? Without it, this proposal is equivalent to sealer/unsealer pairs [0,1] which are quite powerful and well understood. Were this proposal exactly equivalent to sealer/unsealer pairs, I would be on familiar territory.

[0] http://www.erights.org/history/morris73.pdf
[1] http://www.erights.org/elib/capability/ode/ode-capabilities.html#rights-amp

**Comment by erights** [comment]
The README suggests that the motivating use case is to *change* the binding associated with a symbol that a previous call to `ref` had made. Since calls to ref without the second argument necessarily makes a fresh unregistered symbol, if you required that the symbol provided as second argument were already a key in this refCollection, then they would happen to necessarily be unregistered without the API needing to case split on registered vs unregistered symbols.

You just need to prohibit `ref` or two arguments from making a new entry, rather than overriding an existing one.

**Comment by erights** [comment]
Note that I am not saying that even this restricted form of the second argument should be allowed. I'm just fixing one objection to it.

**Comment by rricard** [comment]
This parameter has been added as it was seen as a way to manually associate to an existing ref symbol (and with the nice side effect to be able to name ref symbol). I agree with your remarks that it is probably something we might not want as it creates issues. It is not, to me, an essential part of the proposal that we can easily strip.

**Comment by rricard** [comment]
>  Without it, this proposal is equivalent to sealer/unsealer pairs [0,1] which are quite powerful and well understood. Were this proposal exactly equivalent to sealer/unsealer pairs, I would be on familiar territory.

This is interesting, I'll need to take a look

**Comment by littledan** [comment]
> If ref takes a symbol, then I can pass it a registered symbol (Symbol.for('foo')) or a well-known symbol (Symbol.iterator). These can never be garbage collected, [...]

I don't see why this would be fatal. The lifetime of the entry will be the same as the lifetime of the RefCollection. You can also create this scenario if you hold another symbol alive in some other way. It's probably not so useful to use RefCollection this way, but I don't see why it means that we can't take symbols here in general.

If anyone has gotten the impression that this API lets you replace the value corresponding to a symbol that's already in the RefCollection, this is a misunderstanding: if the symbol is already present, then `ref` would return a new symbol and not use the one provided.

I agree with @rricard that this is not an essential part, and would be fine with removing it if needed. However, the motivation for this support was to help RefCollections fulfill the "template" use case that several TC39 members including @jridgewell has expressed interest in. However, it's possible that these sorts of use cases don't need the "weak" GC semantics, and therefore could use userland solutions instead.

**Comment by ljharb** [comment]
@littledan for the same reasons explained in https://github.com/tc39/ecma262/issues/1194, that prevents Weak Map/Set/Ref from weakly holding onto Symbols.

**Comment by rricard** [comment]
@littledan, as discussed in #10, the "template" use case does not need weak semantics. I think we can safely remove that argument. I'm interested in seeing how we can solve the template use case but it does not look like this proposal is right for that.
